### PR TITLE
Move from `chalk`

### PR DESCRIPTION
**Pull request by Richienb**
Related to https://github.com/chalk/chalk/issues/431

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,10 +11,11 @@
 		"node": ">=12"
 	},
 	"scripts": {
-		"test": "xo && ava"
+		"test": "xo && ava test/index.js && cross-env FORCE_COLOR=0 ava test/no-color.js && cross-env FORCE_COLOR=2 ava test/full-color.js && tsd"
```
Comment (Richienb) [inline package.json]: There is currently no way to confingure ava to set different environment variables across test files.

File: package.json [hunk 2]
```diff
@@ -11,10 +11,11 @@
 		"node": ">=12"
 	},
 	"scripts": {
-		"test": "xo && ava"
+		"test": "xo && ava test/index.js && cross-env FORCE_COLOR=0 ava test/no-color.js && cross-env FORCE_COLOR=3 TERM=dumb ava test/full-color.js && tsd"
```
Comment (Richienb) [inline package.json:14]: `TERM=dumb` is set because it counterintuitively causes `supports-color` to respect `FORCE_COLOR=3`.

**Comment by Richienb** [comment]
@sindresorhus @Qix- Friendly ping :)

File: index.d.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/**
+Use a template string.
+
+@example
+```
+import chalkTemplate from 'chalk';
```
Comment (sindresorhus) [inline index.d.ts]: Incorrect import

**Comment by sindresorhus** [comment]
https://github.com/chalk/chalk/issues/431#issuecomment-950324908

File: test/full-color.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		'\u001B[1mhello {in brackets}\u001B[22m');
+});
+
+test('throw if there is an unclosed block', t => {
+	try {
+		console.log(chalkTemplate`{bold this shouldn't appear ever\}`);
+		t.fail();
+	} catch (error) {
+		t.is(error.message, 'Chalk template literal is missing 1 closing bracket (`}`)');
+	}
```
Comment (sindresorhus) [inline test/full-color.js]: Use `t.throws`

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: test/full-color.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+		+ '\u001B[48;2;144;10;178mthere!\u001B[49m\u001B[22m');
+});
+
+test('properly handle escapes', t => {
+	t.is(chalkTemplate`{bold hello \{in brackets\}}`,
+		'\u001B[1mhello {in brackets}\u001B[22m');
+});
+
+test('throw if there is an unclosed block', t => {
+	t.throws(() => chalkTemplate`{bold this shouldn't work ever\}`, {
```
Comment (sindresorhus) [inline test/full-color.js]:
```suggestion
	t.throws(() => {
		chalkTemplate`{bold this shouldn't work ever\}`;
	}, {
```

For readability. I prefer not using inline arrow function for void functions.

**Comment by sindresorhus** [comment]
You missed https://github.com/chalk/chalk-template/pull/3#issuecomment-955669397

**Comment by sindresorhus** [comment]
^ Chalk should be a dependency.

File: test/full-color.js (hunk 5 above)
Comment (Richienb) [reply test/full-color.js]: After changing it out, it appears that `xo` doesn't like me not assigning the return value to anything, nor voiding it.

File: index.d.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/**
+Use a template string.
```
Comment (sindresorhus) [inline index.d.ts]: This description is not good enough.

File: index.js [hunk 7]
```diff
@@ -125,9 +127,26 @@ export default function template(chalk, temporary) {
...
+
+export default function chalkTemplate(firstString, ...arguments_) {
+	if (!Array.isArray(firstString) || !Array.isArray(firstString.raw)) {
+		// If chalkTemplate() was called by itself or with a string
+		throw new TypeError('A tagged template literal must be provided');
+	}
+
+	const parts = [firstString.raw[0]];
+
+	for (let i = 1; i < firstString.length; i++) {
```
Comment (sindresorhus) [inline index.js]: Nitpick, but `i` => `index`.

File: package.json [hunk 8]
```diff
@@ -1,7 +1,7 @@
 {
 	"name": "chalk-template",
 	"version": "0.0.0",
-	"description": "TODO",
+	"description": "Use tagged template literals with Chalk",
```
Comment (sindresorhus) [inline package.json]: That's not really the purpose though. The purpose is to apply colors using templating syntax.

File: readme.md [hunk 9]
```diff
@@ -1,24 +1,31 @@
...
-> TODO
+> Use [tagged template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals#tagged_templates) with [Chalk](https://github.com/chalk/chalk)
 
 ## Install
 
 ```
 $ npm install chalk-template
 ```
 
+[`chalk`](https://github.com/chalk/chalk) is a required peer dependency.
```
Comment (sindresorhus) [inline readme.md]: Leftover

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by sindresorhus** [comment]
The readme here is missing a lot of stuff from the Chalk readme: https://github.com/chalk/chalk#tagged-template-literal (Also ensure index.d.ts stays in sync).

File: test/full-color.js (hunk 5 above)
Comment (sindresorhus) [reply test/full-color.js]: Can you not do `void chalkTemplate`{bold this shouldn't work ever\}`;`? If not, just use an ESLint ignore comment.

File: package.json [hunk 10]
```diff
@@ -1,7 +1,7 @@
 {
 	"name": "chalk-template",
```
Comment (lorensr) [inline package.json:2]: If you'll be publishing this to npm, this `name` field needs to be changed, unless you have access to the existing npm package named `chalk-template`:

https://www.npmjs.com/package/chalk-template

Comment (lorensr) [reply package.json:2]: Ah, looks like you do! Looking forward 😊

image

**Review by sindresorhus — APPROVED** [review]

**Comment by Qix-** [comment]
Looks fine, there are some things I want to add/change in the future but this is a wonderful foundation for a v1. Thanks @Richienb :)

**Review by Qix- — APPROVED** [review]

**Comment by Qix-** [comment]
By the way, should be documented that this isn't a _drop in_ replacement for Chalk - places where `chalk` is used as a tag for template literals will need to be changed to `chalkTemplate`, whereas normal Chalk usage (e.g. `chalk.red.dim(...)`) remain the same.

Specifically talking about when we release, let's make sure to include this as "migration notes", either here or in chalk v5, preferably both.

**Comment by Richienb** [comment]
> either here or in chalk v5, preferably both.

There are no users of `chalk-template` to inform so just in chalk would be fine.

**Comment by sindresorhus** [comment]
Thanks for doing this, @Richienb 👍

**Pull request merged**